crates/order-validation/src/banned.rs

```diff
@@ -1,22 +1,122 @@
 use {
-    cached::{Cached, TimedCache},
     contracts::ChainalysisOracle,
     ethcontract::{errors::MethodError, futures::future::join_all, H160},
-    std::{collections::HashSet, sync::Mutex},
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+    tokio::sync::RwLock,
 };
 
 /// A list of banned users and an optional registry that can be checked onchain.
 pub struct Users {
     list: HashSet<H160>,
-    onchain: Option<Onchain>,
+    onchain: Option<Arc<Onchain>>,
+}
+
+#[derive(Clone)]
+struct UserMetadata {
+    is_banned: bool,
+    last_updated: Instant,
+}
+
+impl UserMetadata {
+    pub fn with_banned(mut self, banned: bool) -> Self {
+        self.is_banned = banned;
+        self
+    }
+
+    pub fn with_last_updated(mut self, last_updated: Instant) -> Self {
+        self.last_updated = last_updated;
+        self
+    }
 }
 
 struct Onchain {
     contract: ChainalysisOracle,
-    cache: Mutex<TimedCache<H160, bool>>,
+    cache: RwLock<HashMap<H160, UserMetadata>>,
 }
 
-const TTL: u64 = 60 * 60; // 1 hour
+impl Onchain {
+    pub fn new(contract: ChainalysisOracle) -> Arc<Self> {
+        let onchain = Arc::new(Self {
+            contract,
+            cache: Default::default(),
+        });
+
+        onchain.clone().spawn_maintenance_task();
+
+        onchain
+    }
+
+    /// Spawns a background task that periodically checks the cache for expired
+    /// entries and re-run checks for them.
+    ///
+    /// Doesn't clean the cache, so it can grow indefinitely.
+    fn spawn_maintenance_task(self: Arc<Self>) {
+        let cache_expiry = Duration::from_secs(60 * 60);
+        let maintenance_timeout = Duration::from_secs(60);
+        let detector = Arc::clone(&self);
+
+        tokio::task::spawn(async move {
+            loop {
+                let start = Instant::now();
+
+                let expired_data: Vec<_> = {
+                    let now = Instant::now();
+                    let cache = detector.cache.read().await;
+                    cache
+                        .iter()
+                        .filter_map(|(address, metadata)| {
```

Comment:
is `filter_map` necessary here (aren't you simply filtering)?

Reply:
It is required to clone only the filtered items and not the whole collection.